_posts/2015-04-17-Enums-match-mutation-and-moves.md

```diff
@@ -0,0 +1,673 @@
+---
+layout: post
+title: "Mixing matching, mutation, and moves in Rust"
+author: Felix S. Klock II
+description: "A tour of matching and enums in Rust."
+---
+
+One of the primary goals of the Rust project is to enable safe systems
+programming. Systems programming usually implies imperative
+programming, which in turns often implies side-effects, reasoning
+about shared, aliasable state, et cetera.
+
+At the same time, to provide *safety*, Rust programs and data types
+must be structured in a way that allows static checking to ensure
+soundness. Rust has features and restrictions that operate in tandem
+to ease writing programs that can pass these checks and thus ensure
+safety. For example, Rust incorporates the notion of *ownership* deeply
+into the language.
+
+Rust's `match` expression is a construct that offers an interesting
+combination of such features and restrictions. A `match` expression
+takes an input value, classifies it, and then jumps to code written to
+handle the identified class of data.
+
+In this post we explore how Rust processes such data via `match`.
+The crucial elements that `match` and its counterpart `enum` tie
+together are:
+
+* Structural pattern matching of algebraic data types: The `match`
+  construct enables case analysis with ergonomics that are vastly
+  improved over that provided by a C or Java style `switch` statement.
+
+* Exhaustive case analysis, which ensures that no case is omitted
+  when processing an input.
+
+* `match` embraces both imperative and functional styles of
+  programming: The compiler's static analyses work hard to ensure
+  statement-oriented programming remains palatable, rather than
+  forcing everyone to adopt an expression-oriented mindset.
+
+* Destructuring bind of *L-values*: Rust encourages the developer to
```

Comment:
(todo: find less jargony way to write this bullet.  Maybe just "match can borrow or take ownership, as appropriate")